src/charm.py

```diff
@@ -79,29 +126,73 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
             self._open_port_tcp(self._port)
         except subprocess.CalledProcessError:
-            self.unit.status = BlockedStatus("failed to open TCP port for MongoDB")
+            self.unit.status = BlockedStatus(
+                "failed to open TCP port for MongoDB")
             return
 
-        if not self._mongo.is_ready():
+        # check if this unit's deployment of MongoDB is ready
+        if not self._mongo.is_ready(as_stand_alone=True):
+            logger.debug("mongoDB not ready, deferring on start event")
             self.unit.status = WaitingStatus("waiting for MongoDB to start")
             event.defer()
             return
 
-        # initialize replica set
-        self.unit.status = MaintenanceStatus("initialising MongoDB replicaset")
-        logger.debug("initialising replica set")
+        # mongod is now active
+        self.unit.status = ActiveStatus()
+
+        # only leader should initialise the replica set
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self._number_of_expected_units == self._number_of_current_units:
+            logger.debug(
+                "planned units does not match joined units: %s != %s,"
+                + "waiting to initialise replica set",
+                str(self._number_of_expected_units),
+                str(self._number_of_current_units),
+            )
+            return
+
+        # initialise replica set
         try:
-            logger.debug("initialising replica set for the following IPs %s", self._unit_ips)
-            self._mongo.initialize_replica_set(self._unit_ips)
-            self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
-        except (ConnectionFailure, ConfigurationError) as e:
-            logger.error("error initialising replica sets in _on_start: error: %s", str(e))
-            self.unit.status = BlockedStatus("failed to initialise replicaset")
+            self._initialise_replica_set()
+        except (ConnectionFailure, ConfigurationError):
+            self.unit.status = BlockedStatus(
+                "failed to initialise replica set")
+
+    def _on_update_status(self, event):
+        # connect to client for this single unit
+        mongod_unit = self._single_mongo_replica(
+            str(self.model.get_binding(PEER).network.bind_address))
+
+        # if unit is primary then update status
+        if mongod_unit._is_primary:
+            self.unit.status = ActiveStatus("Replica set primary")
+
+    def _on_get_primary_action(self, event):
+        # check if current unit is the primary unit
+        mongod_unit = self._single_mongo_replica(
+            str(self.model.get_binding(PEER).network.bind_address))
+
+        # if unit is primary display this inforamtion and exit
+        if mongod_unit._is_primary:
+            event.set_results({"replica-set-primary": self.unit.name})
             return
```

Comment:
Yes, sorry if this code is confusing, the `README` is correct that "the action can be called in any unit to retrieve the primary". This return on line 184 just prevents the rest of the code from running. ie `# if unit is primary display this inforamtion and exit` and  `# loop through peers and check if one is the primary`

If the current unit isn't the primary, the lines 186-195 loop to check if a peer is a primary, and it if a peer is a primary that is outputted on line 194. Does that clear things up? 

BUT actually, I realize this code can be simplified using the `self.unit_ips` so I will simplify it as so:
```
for unit_ip in self._unit_ips():
            # set up a mongo client for this single replica
            mongod_unit = self._single_mongo_replica(unit_ip)

            # if unit is primary display this inforamtion and exit
            if mongod_unit._is_primary:
                event.set_results({"replica-set-primary": unit.name})
                return
```
thanks for the thorough reviews

Reply:
Sorry Mia, my bad. I incorrectly read the code and thougth  that the loop was other function. The code is clear. Sorry for the confusion.

Reply:
No its really good that you commented because you made me realized I could simplify my code!! I'm super grateful @marceloneppel

Reply:
🤦 Actually I can't do my above solution, since I need access to `unit` to get its name and there is no nice way to get all the units from the application and their IPs.

Reply:
Even with 
```
                
    def _on_get_primary_action(self, event: ops.charm.ActionEvent):
        # get all units
        units = self._peers.units
        units.add(self.unit)

        # check each unit for primary status
        for unit, unit_ip in zip(units, self._unit_ips):
            # set up a mongo client for this single replica
            mongod_unit = self._single_mongo_replica(unit_ip)

            # if unit is primary display this inforamtion and exit
            if mongod_unit._is_primary:
                event.set_results({"replica-set-primary": unit.name})
                return
```
we will run into issues with maintaining the order of the `set` `units`, we could convert the `set` to a `list` but that leads to overhead so I'll keep this solution that I already have